### PR TITLE
[EDA] Extend TDTM to allow disabling via User Profile

### LINKED ISSUE #638: HEDA Idea: Extend HEDA TDTM Handler to allow for the exclusion by Profile (INVESTIGATION)
HEDA Idea: [Change HEDA TDTM Handler to exclude based upon Profile.Name instead of Username
](https://powerofus.force.com/hub-ideas/1391)
Created by: [Corey Snow (Harvard Initiative for Learning and Teaching)](https://powerofus.force.com/_ui/core/userprofile/UserProfilePage?u=0058000000AwHSD&tab=sfdc.ProfilePlatformFeed)

**Idea:** 

Currently, in the TDTM Trigger Handler records, one can specify a list of Users for which the Triggers will not run in Field: hed__Trigger_Handler__c.hed__Usernames_to_Exclude__c
While perhaps somewhat helpful for testing, Usernames are guaranteed to be unique in every Sandbox, which means these values will not survive within the context of normal change management strategies. This is essentially 'hard-coding' of Usernames. Changing to use User Profile (Profile.Name) for this exclusion would cause the logic to persist across Sandboxes in a 'self-healing' way. Exclusions by Profile for all forms of automation constitute a best practice across the core platform (e.g. Validation Rules, WFRs, Process Builder, Flows, Apex, etc)

### PR DESCRIPTION
**Pull request by skosgi1**
# Critical Changes
## Disable Trigger Handlers for Specific User Profiles
We've introduced a new feature to disable TDTM trigger handlers with user profiles. The "Profiles to Exclude" field is automatically added to Trigger Handler in new orgs, but in existing orgs, we recommend that you add the new field to your page layout. See [Display a New Custom Object After a Product Release](https://powerofus.force.com/s/article/EDA-Display-Cust-Obj-Post-Rel).

To access this feature, from App Launcher, find and select Trigger Handler. Select a Trigger Handler to edit and enter one or more user profiles, separating each with a semicolon, in the Profiles to Exclude field. For example, "System Administrator;Marketing User."

If you've created translated versions of user profile names in Translation Workbench, be sure to include the translated version in the Profiles to Exclude field. For a Spanish translation, for example, specify "System Administrator;Administrador del sistema."

See [Disable Trigger Handlers for Specific Users](https://powerofus.force.com/s/article/EDA-Disable-Trigger-Handlers#ariaid-title3).

# Changes

# Issues Closed
Closes #638 

# New Metadata
Profiles_to_Exclude__c
TDTM_ExcludedProfiles
TDTM_ExcludedUserNames

# Deleted Metadata

# Testing Notes
The feature can be tested on any TDTM, below is an example:
1) Goto EDA -> Trigger Handlers -> Select the record for 'TERM_CannotDelete_TDTM'
2) Make sure Set up -> Custom Settings -> Hierarchy Settings -> 'Term Cannot Delete' is set to True 
3) Create a Term with at least one Term Grade child record
4) Delete the record. The record deletion is prevented since it has child records.
5) Add 'System Administrator' (with no quotes) into the field 'Profiles to Exclude'. If you want to add multiple profiles use semi colon Ex: 'System Administrator; Standard User'.
6) Make sure you're system administrator -> Set up -> User -> select your user -> Profile -> 'System Administrator'
6) Now try deleting the Term record (Same as step 4)
7) The term should be deleted

Testing Translations:
1) Repeat Step 1 - 4 from above
2) Add 'Administrador del sistema' to Profiles to Exclude
3) Change your language to Spanish. Set up -> User -> select your user -> language -> Spanish
4) Now try deleting the Term record 
5) The term should be deleted

### CODE REVIEW

File: src/classes/TDTM_ExcludedUserNames.cls [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+public class TDTM_ExcludedUserNames {
+
+    public Static Set<String> userNamesToExclude(TDTM_Global_API.TdtmToken token) {
+
+    	//Get a list of excluded usernames in lowercase
+        Set<String> excludedUserNamesLowercase = new Set<String>();
+        Set<String> excludedUserNames = (token.usernamesToExclude != null ?
+            new Set<String>(token.usernamesToExclude.split('\\s*;\\s*')) : new Set<String>());
+        for (String username : excludedUserNames) {
```
[the remark this answers is not in the document]
Comment (skosgi1) [reply src/classes/TDTM_ExcludedUserNames.cls]: Thanks for the review Ryan, I was not sure if we needed that, but I will take another look and get back to you.

[the remark this answers is not in the document]
Comment (skosgi1) [reply src/classes/TDTM_ExcludedUserNames.cls]: @lifewithryan Looks like when we get current user name from org, we get in lower case example : 'test-1oxzqxoixlk0@example.com' so if the user entered Capital case eg: 'Test-1oxzqxoixlk0@example.com' in 'User Names to Exclude' field then the code converts it to lower case and then compares. However, when we query for Current Profile we don't get lower case for example for sys admin we get 'System Administrator', so we have to have an exact match in 'Profiles to Exclude'. So if we document it such that it needs to have an exact match IMO it works (please let me know your thoughts here). If we want to convert it lower case we may have to do it both for 'Current User Profile' and 'Profiles to Exclude'. please lmk.

File: src/classes/TDTM_Global_API.cls [hunk 2]
```diff
@@ -278,6 +284,39 @@ global class TDTM_Global_API {
...
+        * @param active The token's active flag.
+        * @param async The token's async flag.
+        * @param filterField The token's filter field.
+        * @param filterValue The token's filter value.
+        * @param userManaged The token's user-managed flag.
+        * @param ownedByNamespace The token's owned-by-namespace flag.
+        * @param usernamesToExclude The token's usernames-to-excluded field.
+        * @param profilesToExclude The token's profiles-to-excluded field.
+        *********************************************************************************************************/
+        global TdtmToken(String className, String targetObject, String actions, Decimal loadOrder, Boolean active,
```
[the remark this answers is not in the document]
Comment (skosgi1) [reply src/classes/TDTM_Global_API.cls]: The documentation on global tokens is here https://developer.salesforce.org/EDA/ApexDocumentation/TDTM_Global_API.html, which can be found from here https://powerofus.force.com/s/article/EDA-Deploy-Custom-Apex-TDTM (Read more about this globally exposed class and its methods at ) Not sure if we still maintain the former documentation. FYI @juliesnow

File: src/classes/TDTM_ExcludedUserNames.cls (hunk 1 above)
[the remark this answers is not in the document]
Comment (lifewithryan) [reply src/classes/TDTM_ExcludedUserNames.cls]: @skosgi1 -- I think we are probably okay. If we get people who report it not working and find out it is a case issue we can always come back. For the record I think we could at .toLowerCase() in the ExcludedProfiles class on line 45 and then when we do the contains check on line 165 in TDTM_Config, we could add the toLowerCase() to the currentUserProfile there. No need to do that now, but that would be how I would try to fix it if it becomes an issue.

**Comment by spinkelman** [comment]
Passes QA - Merging
Will close WI

**Pull request merged**